# Reading Files Into Tables

## Read, filter, and sort files

This section includes all the current code for reading, parsing, and filtering LTD's files into the most-convenient dataframe tables for our needs.

When run, you will have the tables `agency_df`, `routes_df`, `stops_df`, `calendar_dates_df`, `trips_df`, and `stop_times_df`.

In [1]:
import pandas as pd

In [3]:
# Constants
springfield_stop_id = "02507"
eugene_stop_id = "02121"
agate_inbound_stop_id = "09965"
agate_outbound_stop_id = "09904"

In [2]:
# Load the GTFS data into dataframes
source_dir = '../gtfs/'
agency_df = pd.read_csv(f'{source_dir}agency.txt')
routes_df = pd.read_csv(f'{source_dir}routes.txt')
stops_df = pd.read_csv(f'{source_dir}stops.txt')
calendar_dates_df = pd.read_csv(f'{source_dir}calendar_dates.txt')
trips_df = pd.read_csv(f'{source_dir}trips.txt')
stop_times_df = pd.read_csv(f'{source_dir}stop_times.txt')

In [3]:
# Trim unnecessary columns
routes_df = routes_df[['route_id', 'route_short_name', 'route_long_name']]
stops_df = stops_df[['stop_id', 'stop_name', 'stop_lat', 'stop_lon', 'parent_station', 'platform_code']]
calendar_dates_df = calendar_dates_df[['service_id', 'date']]
trips_df = trips_df[['route_id', 'service_id', 'trip_id', 'trip_headsign', 'direction_id']]
stop_times_df = stop_times_df[['trip_id', 'departure_time', 'stop_id', 'stop_headsign']]

print (f"Loaded {len(routes_df)} routes, {len(stops_df)} stops, {len(calendar_dates_df)} calendar dates, {len(trips_df)} trips, and {len(stop_times_df)} stop times.")


Loaded 28 routes, 1187 stops, 275 calendar dates, 5240 trips, and 150718 stop times.


In [4]:
# Remove trips with no departure time
stop_times_df = stop_times_df[stop_times_df['departure_time'].notna()]
print("Removing untimed stops yields", len(stop_times_df), "rows.")

Removing untimed stops yields 48235 rows.


Adding route_id and service_id to the stop_times table can be done now in one step. The result will be a larger database table, but will greatly reduce the number of queries after seeding.

In [5]:
#stop_times_df.join(stop_times_df, trips_df, on='trip_id', how='inner')

# Add fields from trips_df to stop_times_df on trip_id
stop_times_df = stop_times_df.merge(trips_df, on='trip_id', how='inner')

In [6]:
stop_times_df.head()

trip_id departure_time stop_id                    stop_headsign route_id   
0   698915       06:31:00   09011  92 EUGENE STATION <> 92 via LCC       92  \
1   698915       06:37:00   09240  92 EUGENE STATION <> 92 via LCC       92   
2   698915       06:49:00   09236  92 EUGENE STATION <> 92 via LCC       92   
3   698915       06:56:00   01806  92 EUGENE STATION <> 92 via LCC       92   
4   698915       07:06:00   02305                92 EUGENE STATION       92   

   service_id      trip_headsign  direction_id  
0    13105100  92 EUGENE STATION             1  
1    13105100  92 EUGENE STATION             1  
2    13105100  92 EUGENE STATION             1  
3    13105100  92 EUGENE STATION             1  
4    13105100  92 EUGENE STATION             1

# File Descriptions

Details about the GTFS file format and the specific fields used by LTD. Includes no-longer-relevant code used when exploring data.

In [ ]:
# Dependencies
import csv

# Parts of this include code used while testing the GTFS data. Data was read using the CSV module,
# so it's not as clean as the Pandas dataframes above.

LTD's GTFS zip file contains the following files:

- agency.txt
- calendar_dates.txt
- fare_attributes.txt
- routes.txt
- shapes.txt
- stop_times.txt
- stops.txt
- trips.txt

### Routes

A listing of all the routes. Contains about 20 records.

The file `routes.txt` is simple and short. We need very little data from it. The main purpose is to show the friendly name to the user while using the route_id in the database.

Routes are non-geographic (ex: The Yellow Line), can branch, and can describe both directions of travel.

- route_id = uniquely identifies and relates to other GTFS tables
- route_type = bus, subway, ferry, etc.

We can ignore these: `agency_id` is always "LTD", `route_desc` is always blank, and `route_type` is always "3".

`route_id` and `route_long_name` all match except for the case of "103" that has a short name of "EmX" (also matches the long name)

### Trips file

5240 entries.

A single instance of a vehicle traveling on a route. Always visits the same stops at the same time of day. No Geographic info. Finer level of detail than routes.txt

- trip_id: Unique identifier
- route_id: Relates to the route it belongs to
- shape_id: index in the shape.txt file.
- service_id: relates to calendar.txt and calendar_dates.txt to describe when it runs
- wheelchair_accessible
- bikes_allowed

In [31]:
# Reading a file with the CSV module and then converting to a Pandas DataFrame
# Update: Pandas can read CSV files directly, so this is not necessary

trips_path = "trips.txt"

# Read CSV file
with open(trips_path, 'r') as file:
    reader = csv.reader(file)
    data = list(reader)

# Convert to Pandas DataFrame
trips_df = pd.DataFrame(data[1:], columns = data[0])

# Select desired columns
# route_id,service_id,trip_id,trip_headsign,block_id,direction_id,shape_id

print(len(trips_df), "records read")
print(trips_df.shape[0], "rows and", trips_df.shape[1], "columns")

print("Unique service IDs:", trips_df["service_id"].nunique())
print("Unique trip IDs:", trips_df["trip_id"].nunique())
print("Unique block IDs:", trips_df["block_id"].nunique())

print ("Service IDs:", trips_df["service_id"].unique())


5240 records read
5240 rows and 7 columns
Unique service IDs: 9
Unique trip IDs: 5240
Unique block IDs: 210
Service IDs: ['13105100' '13105200' '13105300' '13205100' '13205200' '13205300'
 '13105304' '13205304' '13105312']


In [54]:
# Saturday trips
trips_df_sat = trips_df[trips_df["service_id"] == "13205100"]
print(len(trips_df_sat), "Saturday trips")

# Sunday trips
trips_df_sun = trips_df[trips_df["service_id"] == "13205200"]
print(len(trips_df_sun), "Sunday trips")

# Weekday trips
trips_df_mfa = trips_df[trips_df["service_id"] == "13205300"]
# mfb adds route 79x plus 1 extra dropoff for route 36
trips_df_mfb = trips_df[trips_df["service_id"] == "13205304"]
num_weekday_trips = len(trips_df_mfa) + len(trips_df_mfb)
print(f"Weekday trips: {len(trips_df_mfa)} + {len(trips_df_mfb)} = {num_weekday_trips}")

# EmX trips (Route 103)
trips_df_emx = trips_df[trips_df["route_id"] == "103"]
print(len(trips_df_emx), "total EmX trips")

# EmX trips on Saturday
trips_df_emx_sat = trips_df_sat[trips_df_sat["route_id"] == "103"]
print(len(trips_df_emx_sat), "Saturday EmX trips")

# Array of EmX Saturday trip IDs
list_of_sat_emx_trips = trips_df_emx_sat["trip_id"]


791 Saturday trips
611 Sunday trips
Weekday trips: 1134 + 43 = 1177
802 total EmX trips
129 Saturday EmX trips


### Calendar Dates

A many-to-many relationship between calendar dates and service IDs. Service IDs refer to the schedule(s) in use that day.

`calendar_dates.txt`: 276 lines in total. Many dates have 2 entries.

- "service_id": 8-digit number referencing the calendar schedule in use that day. M-F calendars use both 13205300 and 13205304. Sat=13205100. Sun=13205200.
- "date": 8-digit representation of the date as yyyymmdd. It starts on Jan 8 and continues through 20230617. **Does the schedule change next month? YES.**
- "exception_type": 1=added. 2=removed. Always 1 in this file.

The best plan is to look up the current date to find the service_id(s) in use that day. `trips.txt` can then be filtered down to only include trips running that day. This will account for holidays automatically.

Alternatively, since we know the 4 service IDs currently in use and their mapping to days of the week, we can work the other direction and handle holidays manually.


### Stops

Associates a stop name with its ID and its parent station name and bay, if applicable.

This will be useful in the final product for the end user to find a stop name and us to associate it with a stop ID.

`stops.txt`: 1188 lines. 

- `stop_id`: 5 digit code starting at 1. Ending at 99911
- `stop_code`: matches "stop_id". Skip
- `stop_name`: descriptive name. ex: "N/S of Main W of 58th"
- `stop_desc`: Always blank? Drop.
- `stop_lat` and "stop_lon": coords on a map. May be useful.
- `location_type`: always blank. Drop.
- `parent_station` when the stop is a bay, this refers to the station entry.
- `stop_timezone`, `wheelchair_boarding`, and `level_id` always blank. Drop.
- `platform_code`: The bay letter

Stops in a multi-bay station:

- 00704 has a parent station of 99910
- 01040 UO Bay B parent = 99905, platform code "B"
- 01550 UO South parent = 99906. No platform code
- 02101-02121 = Eugene Sta A-U. Parent = 99901
- 02156-02159 = UO C, A, D, E
- 02301-02305 = LCC A-E. Parent = 99904
- 02501-02507 = Spfld A-G, 99902.
- 02510 = Spfld special event. 99902. No bay.
- gateway station parent = 99907 bays A and B
- Santa Clara ABCEF = 99911
- 09927-8 Gateway B, C (spfld or riverbend) = 99907

Last block of numbers: 99901-11 = Eugene, Spfld, Amazon, LCC, UO, UO South, Gateway, Thurston, VRC, RR, SC. All have station type of 1.
Last 7 entries are about arrival zones. First 2 columns match parent station 4th column (stop_desc)
EmX routes don't include Eugene, Spfld, RB stations. Gateway included.

### Stop Times

Defines every time of day a stop is visited. Defines ordered sequence of stop visits. Arrival and departure times can be null.

`stop_times.txt` is the main source of data we are seeking. Sorted by trip_id, not by time or route. Rows are constructed as follows:

- trip_id: 6 digits. Relates to trips.txt file.
- arrival_time and departure_time: 24h format "hh:mm:ss". Blanks for untimed stops. WARNING: some may be after midnight (ex: 24:03:00)
- stop_id: 5 digits. Relates to stops.txt file. (individual stop/bay #s)
- stop_sequence: the stop's position along the route
- stop_headsign: Outside display on bus. ex: "101 EmX EUGENE STATION"
- layover: "True" or "False". Usually False.
- pickup_type, drop_off_type: always blank?
- shape_dist_traveled: Used for rendering partial shapes in a journey planner. Always blank?
- timepoint: 1 if this is a timed stop. 0 if it is an intermediate (untimed stop)

In [44]:
# Process the stop_times.txt file
stop_times_path = "stop_times.txt"

# Read CSV file
with open(stop_times_path, 'r') as file:
    reader = csv.reader(file)
    data = list(reader)

# Convert to Pandas DataFrame
stop_times_df = pd.DataFrame(data[1:], columns = data[0])

# Select desired columns
desired_columns = ["trip_id", "departure_time", "stop_id", "stop_sequence", "stop_headsign"]
stop_times_df = stop_times_df[desired_columns]

print(len(stop_times_df), "records read")
print(stop_times_df.shape[0], "rows and", stop_times_df.shape[1], "columns")

150718 records read
150718 rows and 5 columns


Only keep stops with a timepoint. Departure times from intermediate stops would have to be estimated and add too much complexity to deal with now.

In [4]:
# If a departure time is missing, fill it in with the previous time + 1 second
# DEBUG: This may be failing due to multiple threads. Try running in a single thread.
# Only necessary if we are preparing the full schedule for the database.

# In case the first row is blank:

# previous_time = None

# for index, row in df.iterrows():
#     if pd.isnull(row["departure_time"]):
#         if previous_time is None:
#             # Handle the case of the first row having an empty departure time
#             df.at[index, "departure_time"] = pd.to_datetime("00:00:00", format="%H:%M:%S")
#         else:
#             if previous_time == pd.to_datetime("23:59:59", format="%H:%M:%S"):
#                 next_time = pd.to_datetime("00:00:00", format="%H:%M:%S")
#             else:
#                 next_time = previous_time + pd.Timedelta(seconds=1)
#             df.at[index, "departure_time"] = previous_time + pd.Timedelta(seconds=1)
#     previous_time = row["departure_time"]

# To convert all dates in table back to strings:
# df["departure_time"] = df["departure_time"].dt.strftime("%H:%M:%S")

In [45]:
# Filter out the untimed stops for now. They are difficult to deal with.

stop_times_df = stop_times_df[stop_times_df["departure_time"] != ""]
print("Removing empty times leaves us with", len(stop_times_df), "rows.")

Removing empty times leaves us with 48235 rows.


This was the original strategy to store the times as an actual datetime object in a database. Currently we store it in its original format as a string instead.

- The existing "hh:mm:ss" format is already convenient and uses less space.
- Datetime objects can't natively deal with times after midnight (ex: 24:03:00)
- Datetime adds the complexity of a date, which isn't known.
- Different languages and databases might not interpret it the same.

In [46]:
# Use a custom datetime format to fix hours > 24

def custom_to_datetime(time_string):
    """Convert a time string in the format HH:MM:SS to a datetime.time object."""

    # Skip pd.NaT and empty string values (only occurs when we proccess the entire table)
    if time_string == "":
        return None # or pd.NaT. Allows for further processing of untimed stops.
    if type(time_string) != str:
        print(f"Error: Expected string, got {type(time_string)}")
        return None
    
    nextday = 0
    hours, minutes, seconds = time_string.split(":")
    if int(hours) >= 24:
        hours = str(int(hours) - 24)
        nextday = 1
    newtime = pd.to_datetime(f"{hours}:{minutes}:{seconds}", format="%H:%M:%S")
    return newtime + pd.DateOffset(days=nextday)
    

In [47]:
# Convert departure_time for sorting.

# Whether we do this to the full schedule or just our subset
# will depend on whether we are preparing the full schedule for the database

# OLD: df["departure_time"] = pd.to_datetime(df["departure_time"], format="%H:%M:%S", errors='coerce')

stop_times_df["departure_time"] = stop_times_df["departure_time"].apply(custom_to_datetime)

### Agency Info

`agency.txt`: agency name and timezone info (1 row). We'll keep it to make time zone conversions easier.

### Other files

These files are included by LTD, but we can safely ignore them.

`shapes.txt`: Each route has a GPS shape and is given a "shape_id". In this file, each row has a datapoint with its shape ID, sequence, and GPS coords. To construct a route, you would have to find all by shape_id. 116,136 rows, 25,745 shapes, 1,815 sequences. We can ignore this file of map data. Purely greographical info and optional in GTFS.

`agency.txt` just says that LTD published it on the west coast in English. No useful data

`fare_attributes.txt` lists the prices in dollars and says there are no transfers. No useful data.

`calendar.txt`: LTDs omits this file and uses the calendar_dates.txt file instead. If it were present, we could look up service IDs easily by days of the week.

### Holidays

The holidays published by LTD but not explicitly stored in the files:

- MLK (M Jan 16): M-F
- Presidents (M Jan 16): M-F
- Memorial (M May 29): Sun*
- Juneteenth (M Jun 19): M-F
- Independence (T Jul 4): Sun*
- Labor (M Sep 4): Sun*
- Indigenous (M Oct 9): M-F
- Veterans (Sa, Nov 11): Sat (normal)
- Thanksgiving (Th, Nov 23): No service!
- Day after (F, Nov 24): M-F
- Christmas eve (Su, Dec 24): Sun, but ends at 7:30!
- Christmas day (M, Dec 25): No service!
- New years (M, Jan 1, 2024): Sun*

# Extracting Data From Tables

This section is used for testing different table configurations for best database construction.

In [9]:
# Constants for testing

springfield_stop_id = "02507"
eugene_stop_id = "02121"
agate_inbound_stop_id = "09965"
agate_outbound_stop_id = "09904"

In [10]:
# Describe the data

first_date = calendar_dates_df["date"].min()
last_date = calendar_dates_df["date"].max()
agency_name = agency_df["agency_name"][0]
agency_id = agency_df["agency_id"][0]
agency_timezone = agency_df["agency_timezone"][0]

# Agency data
print(f"Transit information for {agency_name} ({agency_id})")
print(f"Times are in {agency_timezone} time.")

# Calendar data
print ("Calendar data covers the period from", first_date, "to", last_date, "inclusive")
print("-", len(calendar_dates_df["service_id"].unique()), "unique service IDs (schedules), and",
    len(calendar_dates_df["date"].unique()), "unique dates"
)

# Routes and trips data
print("-", len(routes_df), "routes, and", len(trips_df), "trips")

# Stop times data
print("-", len(stop_times_df), "stop_times")

# Filter rows based on "stop_id" column
# stop_df = stop_times_df.loc[stop_times_df["stop_id"] == springfield_stop_id]
# num_rows_after = stop_df.shape[0]

# print(f"Number of rows after filtering specific stop: {num_rows_after}")

Transit information for Lane Transit District (LTD)
Times are in America/Los_Angeles time.
Calendar data covers the period from 20230108 to 20230617 inclusive
- 10 unique service IDs (schedules), and 161 unique dates
- 28 routes, and 5240 trips
- 48235 stop_times


### Utility functions

In [46]:
# Get service IDs for a specific date
def get_service_ids_by_date(date):
    """Return a list of service IDs for a specific date in the format YYYYMMDD"""
    schedule_df = calendar_dates_df.loc[calendar_dates_df["date"] == int(date)]
    return schedule_df["service_id"].tolist()

In [59]:
# Get departure times for a specific stop on a specific date
def get_departure_times_by_stop_and_date(stop_id, date):
    """Return a list of departure times for a specific stop id (string) on a specific date in the format YYYYMMDD"""
    stop_id = str(stop_id) # TODO: Pad left with zeroes to make 5 digits
    service_ids = get_service_ids_by_date(date)
    print(f"Found {len(service_ids)} service IDs for date {date}")
    stop_df = stop_times_df.loc[stop_times_df["stop_id"] == stop_id]
    print(f"Found {stop_df.shape[0]} rows for stop {stop_id}")
    stop_df = stop_df.loc[stop_df["service_id"].isin(service_ids)]
    print(f"Found {stop_df.shape[0]} rows for stop {stop_id} and date {date}")
    departure_times = stop_df["departure_time"].tolist()
    departure_times.sort()
    return departure_times


In [62]:
# Get entries for a specific stop on a specific date
def get_entries_by_stop_and_date(stop_id, date):
    """Return data frame for a specific stop id (string) on a specific date in the format YYYYMMDD"""
    stop_id = str(stop_id) # TODO: Pad left with zeroes to make 5 digits
    service_ids = get_service_ids_by_date(date)
    print(f"Found {len(service_ids)} service IDs for date {date}")
    stop_df = stop_times_df.loc[stop_times_df["stop_id"] == stop_id]
    print(f"Found {stop_df.shape[0]} rows for stop {stop_id}")
    stop_df = stop_df.loc[stop_df["service_id"].isin(service_ids)]
    print(f"Found {stop_df.shape[0]} rows for stop {stop_id} and date {date}")
    stop_df.sort_values(by=["departure_time"], inplace=True)
    return stop_df

In [63]:
# Test Sunday, June 4, 2023
get_entries_by_stop_and_date(springfield_stop_id, 20230604)

Found 1 service IDs for date 20230604
Found 400 rows for stop 02507
Found 52 rows for stop 02507 and date 20230604


trip_id departure_time stop_id           stop_headsign route_id   
32811   703482       08:10:00   02507  101 EmX EUGENE STATION      103  \
33314   703507       08:25:00   02507  101 EmX EUGENE STATION      103   
34571   703625       08:40:00   02507  101 EmX EUGENE STATION      103   
35634   703702       08:55:00   02507  101 EmX EUGENE STATION      103   
36453   703730       09:08:00   02507  101 EmX EUGENE STATION      103   
34126   703602       09:24:00   02507  101 EmX EUGENE STATION      103   
35116   703664       09:39:00   02507  101 EmX EUGENE STATION      103   
36048   703714       09:55:00   02507  101 EmX EUGENE STATION      103   
32880   703484       10:10:00   02507  101 EmX EUGENE STATION      103   
37009   703774       10:25:00   02507  101 EmX EUGENE STATION      103   
33383   703509       10:40:00   02507  101 EmX EUGENE STATION      103   
34640   703627       10:55:00   02507  101 EmX EUGENE STATION      103   
35703   703704       11:10:00   02507  101 EmX EUGENE STATION      103   
36522   703732       11:25:00   02507  101 EmX EUGENE STATION      103   
34195   703604       11:40:00   02507  101 EmX EUGENE STATION      103   
35185   703666       11:56:00   02507  101 EmX EUGENE STATION      103   
36117   703716       12:11:00   02507  101 EmX EUGENE STATION      103   
32949   703486       12:26:00   02507  101 EmX EUGENE STATION      103   
37078   703776       12:41:00   02507  101 EmX EUGENE STATION      103   
33452   703511       12:56:00   02507  101 EmX EUGENE STATION      103   
34709   703629       13:11:00   02507  101 EmX EUGENE STATION      103   
35772   703706       13:26:00   02507  101 EmX EUGENE STATION      103   
36591   703734       13:41:00   02507  101 EmX EUGENE STATION      103   
34264   703606       13:56:00   02507  101 EmX EUGENE STATION      103   
35254   703668       14:11:00   02507  101 EmX EUGENE STATION      103   
36186   703718       14:26:00   02507  101 EmX EUGENE STATION      103   
33018   703488       14:41:00   02507  101 EmX EUGENE STATION      103   
37147   703778       14:56:00   02507  101 EmX EUGENE STATION      103   
33521   703513       15:11:00   02507  101 EmX EUGENE STATION      103   
34778   703631       15:26:00   02507  101 EmX EUGENE STATION      103   
35841   703708       15:41:00   02507  101 EmX EUGENE STATION      103   
36660   703736       15:56:00   02507  101 EmX EUGENE STATION      103   
34333   703608       16:11:00   02507  101 EmX EUGENE STATION      103   
35323   703670       16:26:00   02507  101 EmX EUGENE STATION      103   
36255   703720       16:41:00   02507  101 EmX EUGENE STATION      103   
33087   703490       16:56:00   02507  101 EmX EUGENE STATION      103   
37216   703780       17:11:00   02507  101 EmX EUGENE STATION      103   
33590   703515       17:26:00   02507  101 EmX EUGENE STATION      103   
34847   703633       17:41:00   02507  101 EmX EUGENE STATION      103   
35910   703710       17:56:00   02507  101 EmX EUGENE STATION      103   
36729   703738       18:11:00   02507  101 EmX EUGENE STATION      103   
34402   703610       18:26:00   02507  101 EmX EUGENE STATION      103   
35392   703672       18:41:00   02507  101 EmX EUGENE STATION      103   
36324   703722       18:55:00   02507  101 EmX EUGENE STATION      103   
33156   703492       19:10:00   02507  101 EmX EUGENE STATION      103   
37285   703782       19:25:00   02507  101 EmX EUGENE STATION      103   
33659   703517       19:40:00   02507  101 EmX EUGENE STATION      103   
34916   703635       19:55:00   02507  101 EmX EUGENE STATION      103   
35979   703712       20:09:00   02507  101 EmX EUGENE STATION      103   
36798   703740       20:24:00   02507  101 EmX EUGENE STATION      103   
34471   703612       20:39:00   02507  101 EmX EUGENE STATION      103   
35461   703674       20:54:00   02507  101 EmX EUGENE STATION      103   

       service_id                          trip_headsign  direction_id  
32811    13

In [65]:
# Test Monday, June 5, 2023
get_entries_by_stop_and_date(springfield_stop_id, 20230605)

Found 2 service IDs for date 20230605
Found 400 rows for stop 02507
Found 75 rows for stop 02507 and date 20230605


trip_id departure_time stop_id           stop_headsign route_id   
38894   704373       06:12:00   02507  101 EmX EUGENE STATION      103  \
40106   704527       06:21:00   02507  101 EmX EUGENE STATION      103   
40732   704560       06:41:00   02507  101 EmX EUGENE STATION      103   
38637   704338       06:55:00   02507  101 EmX EUGENE STATION      103   
42784   704712       07:06:00   02507  101 EmX EUGENE STATION      103   
...        ...            ...     ...                     ...      ...   
43965   704754       21:55:00   02507  101 EmX EUGENE STATION      103   
45759   704965       22:09:00   02507  101 EmX EUGENE STATION      103   
41215   704574       22:25:00   02507  101 EmX EUGENE STATION      103   
43482   704740       22:54:00   02507  101 EmX EUGENE STATION      103   
40630   704543       23:24:00   02507  101 EmX EUGENE STATION      103   

       service_id                          trip_headsign  direction_id  
38894    13205300  103 EmX WEST 11TH <> COMMERCE STATION             1  
40106    13205300                 101 EmX EUGENE STATION             1  
40732    13205300  103 EmX WEST 11TH <> COMMERCE STATION             1  
38637    13205300  103 EmX WEST 11TH <> COMMERCE STATION             1  
42784    13205300  103 EmX WEST 11TH <> COMMERCE STATION             1  
...           ...                                    ...           ...  
43965    13205300                 101 EmX EUGENE STATION             1  
45759    13205300  103 EmX WEST 11TH <> COMMERCE STATION             1  
41215    13205300                 101 EmX EUGENE STATION             1  
43482    13205300  103 EmX WEST 11TH <> COMMERCE STATION             1  
40630    13205300                 101 EmX EUGENE STATION             1  

[75 rows x 8 columns]

# Create Databases

This section is constructed in 2 parts:

1. Create a sqlite database file from the above tables. Can be used by offline applications or for testing without a database.
2. Load data into postgres using `pgloader`

## Creating the local SQLite database

### Only add a new database file if you need one

In [11]:
# Create a new database
from pathlib import Path
Path('my_data.db').touch()

### Connect to the database

In [7]:
# Connect to the database
import sqlite3
conn = sqlite3.connect('my_data.db')
c = conn.cursor()

### Create Tables

You can run commands to create tables manually, but pandas just overwrites them anyway, so we'll skip that part.

We've already parsed out all our unwanted data in the first section, so now we can just write the tables with pandas.

 We have 6 tables: `agency_df`, `routes_df`, `stops_df`, `calendar_dates_df`, `trips_df`, and `stop_times_df`.

 TODO: Figure out how to describe table configuration to pandas (ex: indexes, key fields, data types, etc.)

In [8]:
# Create tables

# We have agency_df, routes_df, stops_df,
# calendar_dates_df, trips_df, and stop_times_df
agency_df.to_sql('agency', conn, if_exists='replace')
routes_df.to_sql('routes', conn, if_exists='replace')
stops_df.to_sql('stops', conn, if_exists='replace')
calendar_dates_df.to_sql('calendar_dates', conn, if_exists='replace')
trips_df.to_sql('trips', conn, if_exists='replace')
stop_times_df.to_sql('stop_times', conn, if_exists='replace')


48235

### Close the database

Clean up loose ends when we finish

In [69]:
conn.close()

## Load data from SQLite into PostgreSQL

If you haven't already done so, install [pgloader](https://pgloader.readthedocs.io/).

```bash
sudo apt-get update && sudo apt-get install pgloader
```

Then the one-line command to perform the migration is:

```bash
pgloader ./my_data.db postgresql://pguser:pgpass@pgserver/dbname

# For my local installation, it looked like this (not counting the password):
pgloader ./my_data.db postgresql://postgres:pgpass@localhost/ltd
```